students/uss_tomasz/lesson_12_debugging/12_3_valdated_user_base/validate_user_base.py

```diff
@@ -0,0 +1,120 @@
+    parser.add_argument('-z', action='store', dest='zip_code', default='00-000', help='Zip code of user.')
+    received_args = parser.parse_args(args)
+    if received_args.email is None or received_args.password is None:
+        return {}
+    else:
+        results = {'email': received_args.email, 'password': received_args.password,
+                   'telephone': received_args.telephone, 'zip_code': received_args.zip_code}
+        return results
+
+
+def is_record_correct(record_type, record):
+    return re.match(REGEX_DATABASE[record_type], record) is not None
+
+
+def validate_registration(registration):
+    for position in registration:
+        if not is_record_correct(position, registration[position]):
+            return False
+    return True
+
+
+def csv_existence_checker(csv_file_name):
+    if not os.path.isfile(csv_file_name):
+        with open(csv_file_name, 'w', newline='') as csv_file:
+            writer = csv.writer(csv_file)
+            writer.writerow(['email', 'password', 'telephone', 'zip-code'])  # columns names
+
+
+def get_csv_content(csv_filename):
+    csv_content = []
+    with open(csv_filename) as csv_file:
+        csv_reader = csv.reader(csv_file, delimiter=',')
+        for row in csv_reader:
+            csv_content.append(row)
+    return csv_content
+
+
+def is_registration_inside_csv(csv_file_name, registration):
+    csv_content = get_csv_content(csv_file_name)
+    line_count = 0
+    for row in csv_content:
+        if line_count == 0:
+            pass  # columns names
+        else:
+            if row[0] == registration['email']:
+                logging.log(logging.WARNING, "This registration is already inside file.")
+                return True
+        line_count += 1
+    logging.log(logging.INFO, "This registration is new inside CSV file.")
+    return False
+
+
+def csv_entry(csv_file_name, registration):
+    with open(csv_file_name, 'a', newline='') as csv_file:
+        writer = csv.writer(csv_file)
+        writer.writerow([registration['email'], registration['password'], registration['telephone'],
+                         registration['zip_code']])
+
+
+def execute_script(args, csv_file_name):
+    csv_existence_checker(csv_file_name)
+    user_input = parse_args(args)
+    if len(user_input) == 0:
+        logging.log(logging.WARNING, "Received parameters are missing")
+    else:
+        logging.log(logging.INFO, "Received following registration:")
+        logging.log(logging.INFO, user_input)
+        if not validate_registration(user_input):
+            logging.log(logging.WARNING, "This registration contains wrong record.")
+        else:
+            logging.log(logging.INFO, "This registration is correctly validated by regular expressions.")
+            if not is_registration_inside_csv(csv_file_name, user_input):
+                csv_entry(csv_file_name, user_input)
```

Comment:
Although from algorithmical point of view this is correct, it's easier to read if you avoid nested checks. So the code could be structured like:
```
if not user_input:
    logging.log(logging.WARNING, "Received parameters are missing")
    return
logging.log(logging.INFO, "Received following registration:")
logging.log(logging.INFO, user_input)
if not validate_registration(user_input):
    logging.log(logging.WARNING, "This registration contains wrong record.")
    return
logging.log(logging.INFO, "This registration is correctly validated by regular expressions.")
if is_registration_inside_csv(csv_file_name, user_input):
    logging.log(logging.INFO, "Entry already exists.")
    return
csv_entry(csv_file_name, user_input)
```

This way we check all preconditions upfront keeping linear program flow and drop further execution if any fails.